tests/unit/test_compress.py

```diff
@@ -0,0 +1,382 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import gzip
+import io
+from copy import deepcopy
+
+import pytest
+
+from botocore.compat import urlencode
+from botocore.compress import COMPRESSION_MAPPING, maybe_compress_request
+from botocore.config import Config
+from tests import mock
+
+
+def _make_op(
+    request_compression=None,
+    has_streaming_input=False,
+    streaming_metadata=None,
+):
+    op = mock.Mock()
+    op.request_compression = request_compression
+    op.has_streaming_input = has_streaming_input
+    if streaming_metadata is not None:
+        streaming_shape = mock.Mock()
+        streaming_shape.metadata = streaming_metadata
+        op.get_streaming_input.return_value = streaming_shape
+    return op
+
+
+OP_NO_COMPRESSION = _make_op()
+OP_WITH_COMPRESSION = _make_op({'encodings': ['gzip']})
+OP_UNKNOWN_COMPRESSION = _make_op({'encodings': ['foo']})
+OP_MULTIPLE_COMPRESSIONS = _make_op({'encodings': ['gzip', 'foo']})
+STREAMING_OP_WITH_COMPRESSION = _make_op(
+    {'encodings': ['gzip']},
+    True,
+    {},
+)
+STREAMING_OP_WITH_COMPRESSION_REQUIRES_LENGTH = _make_op(
+    {'encodings': ['gzip']},
+    True,
+    {'requiresLength': True},
+)
+REQUEST_BODY = (
+    b'Action=PutMetricData&Version=2010-08-01&Namespace=Namespace'
+    b'&MetricData.member.1.MetricName=metric&MetricData.member.1.Unit=Bytes'
+    b'&MetricData.member.1.Value=128'
+)
+
+REQUEST_BODY_STRING = REQUEST_BODY.decode('utf-8')
+
+DEFAULT_COMPRESSION_CONFIG = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=10420,
+)
+COMPRESSION_CONFIG_128_BYTES = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=128,
+)
+COMPRESSION_CONFIG_1_BYTE = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=1,
+)
+
+
+class NonSeekableStream:
+    def __init__(self, buffer):
+        self._buffer = buffer
+
+    def read(self, size=None):
+        return self._buffer.read(size)
+
+
+def _request_dict(body=None, headers=None):
+    if body is None:
+        body = b''
```

Comment:
The reason we often use `None` over a more reasonable default is because you can't use an empty list or dict in function signatures. Otherwise, it's evaluated at runtime before any code is executed and assigned to a singleton that carries state between invocations.

e.g.
```python
def test(i, data={}):
   data[i] = 'hello'
   print(data)

for i in range(5):
    test(i)

{0: 'hello'}
{0: 'hello', 1: 'hello'}
{0: 'hello', 1: 'hello', 2: 'hello'}
{0: 'hello', 1: 'hello', 2: 'hello', 3: 'hello'}
{0: 'hello', 1: 'hello', 2: 'hello', 3: 'hello', 4: 'hello'}
```

If we're just declaring something like an empty byte-string we can just do:
`def _request_dict(body=b"", headers=None):`

Reply:
Ah yeah this was a mistake. I should know that its only mutable defaults that are concerning. I'll fix.